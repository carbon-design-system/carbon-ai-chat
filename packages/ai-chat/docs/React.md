---
title: Using with React
---

### Overview

Carbon AI Chat exports two React components.

If you want to use the `float` layout, use {@link ChatContainer}. Use the {@link ChatCustomElement} for custom sizes, such as a sidebar, full screen, or nested in your UI.

**Currently, this component does not support SSR, so if you are using Next.js or similar frameworks, make sure you render this component in client only modes.**

For more information, see [the examples page for more examples](https://github.com/carbon-design-system/carbon-ai-chat/tree/main/examples/react).

### Installation

Install by using `npm`:

```bash
npm install @carbon/ai-chat
```

Or using `yarn`:

```bash
yarn add @carbon/ai-chat
```

_Be sure to check for required peerDependencies._

#### Basic example

Render this component in your application and provide the configuration options for the Carbon AI Chat as a prop. Refer to the following example.

```javascript
import React from "react";
import { ChatContainer } from "@carbon/ai-chat";

const chatOptions = {
  // Your configuration object.
};

function App() {
  return <ChatContainer config={chatOptions} />;
}
```

### Using ChatContainer

The {@link ChatContainer} is a functional component that loads and renders an instance of the Carbon AI Chat when it mounts, and deletes the instance when unmounted. If the configuration for the Carbon AI Chat changes, it also deletes the previous Carbon AI Chat and creates a new one with the new configuration. It can also manage React portals for user-defined responses.

See {@link ChatContainerProps} for an explanation of the various accepted props.

### Using ChatCustomElement

This library provides the {@link ChatCustomElement} component, which can be used to render the Carbon AI Chat inside a custom element. Use it if you want to change the location where the Carbon AI Chat renders. This component renders an element in your React app and uses that element as the custom element for rendering the Carbon AI Chat. See {@link ChatCustomElementProps} for an explanation of the various accepted props.

This component's default behavior adds and removes a class from the main window of the Carbon AI Chat. It also applies the same behavior to your custom element to manage the visibility of the Carbon AI Chat when it opens or closes. When the Carbon AI Chat closes, it adds a classname to the Carbon AI Chat main window to hide the element. Your custom element receives another classname to set its width and height to 0, so that it doesn't take up space.

**Note:** In the use case where you are using a custom element but also using the Carbon AI Chat's native launcher, the custom element must remain visible as it also contains the launcher. With that in mind, you should really provide your own launcher.

If you don't want these behaviors, provide your own `onViewChange` prop to {@link ChatCustomElementProps.onViewChange} and provide your logic for controlling the visibility of the Carbon AI Chat. If you want custom animations when the Carbon AI Chat opens and closes, this is the mechanism to do that. Refer to the following example.

See {@link ChatCustomElementProps} for an explanation of the various accepted props.

```javascript
import React from "react";
import { ChatCustomElement } from "@carbon/ai-chat";

import "./App.css";

const chatOptions = {
  /* Carbon AI Chat options */
};

function App() {
  return <ChatCustomElement className="MyCustomElement" config={chatOptions} />;
}
```

```css
.MyCustomElement {
  position: absolute;
  left: 100px;
  top: 100px;
  width: 500px;
  height: 500px;
}
```

### Config object changes

The chat is not capable of handling in-place changes to the config object. If any of the properties in the config object are changed, then the existing chat will be discarded and a whole new chat will be created using the new config properties. A deep equal comparison of the config object is done to detect changes. Note that DOM elements and functions are compared using a strict `===` comparison.

Of note, the above means that if you are creating a new callback function for something like {@link PublicConfigMessaging.customSendMessage} on each render of your component, this would cause the chat to be discarded and recreated each time your component is rendered.

Below is an example of bad code where a new {@link PublicConfigMessaging.customSendMessage} function causes the chat to get recreated each time `App` is rendered.

#### Wrong example

```javascript
// DO NOT DO THIS!

function App() {
  // This is bad because it creates a new customSendMessage function every time App is rendered which will cause the
  // ChatContainer to create a new chat every time.
  const customSendMessage = (message: MessageRequest) => {
    console.log("Sending message", message);
  };

  const config: PublicConfig = { messaging: { customSendMessage } };

  return <ChatContainer config={config} />;
}
```

The quickest solution to this is to move the config object outside of your render function.

#### Correct example #1

```javascript
const customSendMessage = (message: MessageRequest) => {
  console.log("Sending message", message);
};

const config: PublicConfig = { messaging: { customSendMessage } };

function App() {
  return <ChatContainer config={config} />;
}
```

However this does not work if your {@link PublicConfigMessaging.customSendMessage} function requires access to state or props from your application. If you need that, you can use `useCallback` to ensure only a single instance is created from the dependencies. Note in this example that if the instance of the `messageService` prop changes, the whole chat will get recreated. If you don't expect that to happen, this will work for you. Otherwise, you may need to use a ref.

#### Correct example #2

```javascript
function App({ messageService }: any) {
  // Creates a new customSendMessage for each instance of messageService.
  const customSendMessage = useCallback(
    (message: MessageRequest) => {
      messageService(message);
    },
    [messageService] // If messageService changes, the chat will be recreated.
  );

  const config: PublicConfig = { messaging: { customSendMessage } };

  return <ChatContainer config={config} />;
}
```

If you have dependencies that shouldn't recreate the chat, you can store them in a ref. In this example, the chat will not be recreated.

#### Correct example #3

```javascript
function App({ messageService }: any) {
  const messageServiceRef = useRef(messageService);
  messageServiceRef.current = messageService;

  const customSendMessage = useCallback((message: MessageRequest) => {
    messageServiceRef.current(message);
  }, []);

  const config: PublicConfig = { messaging: { customSendMessage } };

  return <ChatContainer config={config} />;
}
```

### Accessing instance methods

You can use the {@link ChatContainerProps.onBeforeRender} or {@link ChatContainerProps.onAfterRender} props to access the Carbon AI Chat's instance if you need to call instance methods later. This example renders a button that toggles the Carbon AI Chat open and only renders after the instance becomes available. Refer to the following example.

```javascript
import React, { useCallback, useState } from "react";
import { ChatContainer } from "@carbon/ai-chat";

const chatOptions = {
  // Your configuration object.
};

function App() {
  const [instance, setInstance] = useState(null);

  const toggleWebChat = useCallback(() => {
    instance.toggleOpen();
  }, [instance]);

  function onBeforeRender(instance) {
    // Make the instance available to the React components.
    setInstance(instance);
  }

  return (
    <>
      {instance && (
        <button type="button" onClick={toggleWebChat}>
          Toggle Carbon AI Chat
        </button>
      )}
      <ChatContainer config={chatOptions} onBeforeRender={onBeforeRender} />
    </>
  );
}
```

### User-defined responses

This component can also manage `user_defined` responses. (See {@link UserDefinedItem}). You must pass a {@link ChatContainerProps.renderUserDefinedResponse} function as a render prop. This function returns a React component that renders content for the specific message that relates to that response. Be sure to review [UI customization](Customization.md).

Treat the {@link ChatContainerProps.renderUserDefinedResponse} prop like any typical React render prop. It is called every time the App rerenders and every time a new `user_defined` message is received. This means you don't want to be calling functions from inside {@link ChatContainerProps.renderUserDefinedResponse} that you don't want called on every render. Consider putting those function calls inside the React component you render with a `useEffect`.

Refer to the following example.

```javascript
import React from 'react';
import { ChatContainer } from '@carbon/ai-chat';
import { Chart } from './Chart';
import { UserDefinedResponseExample } from './Example';

const chatOptions = {
  // Your configuration object.
};

function App() {
  return <ChatContainer renderUserDefinedResponse={renderUserDefinedResponse} config={chatOptions} />;
}

function someFunction() {}

function renderUserDefinedResponse(state, instance) {
  const { messageItem } = state;
  // The event here contains details for each user defined response that needs to be rendered.
  // You can also pass information from your components props or state into the component your are returning.
  if (messageItem) {
    switch (messageItem.user_defined?.user_defined_type) {
      case 'chart':
        someFunction(); // If you do this, this function will get called on every render!
        return (
          <div className="padding">
            {/* Instead, pass someFunction as a prop and run it when the component first mounts with a useEffect(() => { someFunction() }, []). If you are using Strict mode in developement, refer to https://react.dev/learn/synchronizing-with-effects#how-to-handle-the-effect-firing-twice-in-development  */}
            <Chart content={messageItem.user_defined.chart_data as string} onMount={someFunction}/>
          </div>
        );
      case 'green':
        return <UserDefinedResponseExample text={messageItem.user_defined.text as string} />;
      default:
        return null;
    }
    // We are just going to show a skeleton state here if we are waiting for a stream, but you can instead have another
    // switch statement here that does something more specific depending on the component.
    return <AISkeletonPlaceholder className="fullSkeleton" />;
  }
  return null;
}

```

You may also want your `user_defined` responses to inherit props from your application state. In that case, you will want to bring {@link ChatContainerProps.renderUserDefinedResponse} into your component and wrap it in `useCallback`.

```javascript
import React, { useCallback, useEffect, useState } from 'react';
import { ChatContainer } from '@carbon/ai-chat';

const chatOptions = {
  // Your configuration object.
};

function App() {

  const [stateText, setStateText] = useState<string>('Initial text');

  useEffect(() => {
    // This just updates the stateText every two seconds with Date.now()
    setInterval(() => setStateText(Date.now().toString()), 2000);
  }, []);

  const renderUserDefinedResponse = useCallback(
    (state: RenderUserDefinedState, instance: ChatInstance) => {
      const { messageItem } = state;
      // The event here will contain details for each user defined response that needs to be rendered.
      if (messageItem) {
        switch (messageItem.user_defined?.user_defined_type) {
          case 'green':
            // Pass in the new state as a prop!
            return (
              <UserDefinedResponseExample text={messageItem.user_defined.text as string} parentStateText={stateText} />
            );
          default:
            return null;
        }
      }

      // We are just going to show a skeleton state here if we are waiting for a stream, but you can instead have another
      // switch statement here that does something more specific depending on the component.
      return <AISkeletonPlaceholder className="fullSkeleton" />;
    },
    [stateText], // Only update if stateText changes.
  );

  return <ChatContainer renderUserDefinedResponse={renderUserDefinedResponse} config={chatOptions} />;
}
```

### Writable Elements

This component also has several elements inside the chat that you can add extra content to with a writeable element. The {@link ChatContainerProps.renderWriteableElements} prop is an object with the key as the area you want to render a component to and the value being the component to render there. Be sure to review [UI customization](Customization.md).

Similarly to the {@link ChatContainerProps.renderUserDefinedResponse} prop, if you define your {@link ChatContainerProps.renderWriteableElements} object inside your component, it will be re-created every time your component renders. To avoid this, consider wrapping `renderWriteableElements` in `useMemo` or defining it outside your component. When wrapping with `useMemo` you can also pass values from your component into the writeable elements.

Refer to the following example.

```javascript
import React, { useMemo, useState } from "react";
import { ChatContainer } from "@carbon/ai-chat";
import { AIExplanationTooltipContent } from "./AIExplanationTooltipContent";

const chatOptions = {
  // Your configuration object.
};

function App() {
  const [modelsInUse, setModelsInUse] = useState(["granite-13b-instruct-v2"]);

  const renderWriteableElements = useMemo(
    () => ({
      aiTooltipAfterDescriptionElement: (
        <AIExplanationTooltipContent
          location="aiTooltipAfterDescriptionElement"
          modelsUsed={modelsInUse}
        />
      ),
    }),
    [modelsInUse]
  );

  return (
    <ChatContainer
      renderWriteableElements={renderWriteableElements}
      config={chatOptions}
    />
  );
}
```

### Testing with Jest

Carbon AI Chat exports as an ES module and does not include a CJS build. Please refer to the [Jest documentation](https://jestjs.io/docs/code-transformation) for information about transforming ESM to CJS for Jest using `babel-jest` or `ts-jest`.

You may need to add configuration similar to the following to your Jest configuration.

```
"transform": {
  "\\.[jt]sx?$": "babel-jest",
},
{
  transformIgnorePatterns: [
    '/node_modules/(?!(@?lit.*|@carbon/.+)/)'
  ],
}
```
