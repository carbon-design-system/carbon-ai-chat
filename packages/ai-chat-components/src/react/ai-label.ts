/*
 *  Copyright IBM Corp. 2025
 *
 *  This source code is licensed under the Apache-2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  @license
 */

import { createComponent } from "@lit/react";
import React from "react";

// Export the actual class for the component that will *directly* be wrapped with React.
import CarbonAILabelElement from "@carbon/web-components/es/components/ai-label/ai-label.js";

const AILabel = createComponent({
  tagName: "cds-ai-label",
  elementClass: CarbonAILabelElement,
  react: React,
});

export default AILabel;
