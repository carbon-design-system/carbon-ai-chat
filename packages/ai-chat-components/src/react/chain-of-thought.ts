/*
 *  Copyright IBM Corp. 2025
 *
 *  This source code is licensed under the Apache-2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  @license
 */

import { createComponent } from "@lit/react";
import React from "react";

import CDSChatChainOfThoughtElement from "../components/chain-of-thought/src/cds-aichat-chain-of-thought.js";
import {
  type ChainOfThoughtOnToggle,
  type ChainOfThoughtStep,
  ChainOfThoughtStepStatus,
} from "../components/chain-of-thought/src/types.js";

const ChainOfThought = createComponent({
  tagName: "cds-aichat-chain-of-thought",
  elementClass: CDSChatChainOfThoughtElement,
  react: React,
});

export type { ChainOfThoughtOnToggle, ChainOfThoughtStep };
export { ChainOfThoughtStepStatus };
export default ChainOfThought;
