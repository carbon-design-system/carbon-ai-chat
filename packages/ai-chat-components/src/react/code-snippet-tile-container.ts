/*
 *  Copyright IBM Corp. 2025
 *
 *  This source code is licensed under the Apache-2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  @license
 */

import { createComponent } from "@lit/react";
import React from "react";

// Export the actual class for the component that will *directly* be wrapped with React.
import CDSAIChatCodeSnippetTileContainer from "../components/code-snippet/src/code-snippet-tile-container.js";

const CodeSnippetTileContainer = createComponent({
  tagName: "cds-aichat-code-snippet-tile-container",
  elementClass: CDSAIChatCodeSnippetTileContainer,
  react: React,
  events: {
    onChange: "content-change",
  },
});

export default CodeSnippetTileContainer;
