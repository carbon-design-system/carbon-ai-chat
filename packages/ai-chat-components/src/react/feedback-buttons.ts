/*
 *  Copyright IBM Corp. 2025
 *
 *  This source code is licensed under the Apache-2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  @license
 */

import { createComponent } from "@lit/react";
import React from "react";

import CDSChatFeedbackButtonsElement from "../components/feedback-buttons/src/cds-aichat-feedback-buttons.js";

const FeedbackButtons = createComponent({
  tagName: "cds-aichat-feedback-buttons",
  elementClass: CDSChatFeedbackButtonsElement,
  react: React,
});

export default FeedbackButtons;
