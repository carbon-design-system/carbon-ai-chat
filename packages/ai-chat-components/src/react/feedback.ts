/*
 *  Copyright IBM Corp. 2025
 *
 *  This source code is licensed under the Apache-2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  @license
 */

import { createComponent } from "@lit/react";
import React from "react";

import CDSChatFeedbackElement from "../components/feedback/src/cds-aichat-feedback.js";
import {
  type FeedbackInitialValues,
  type FeedbackSubmitDetails,
} from "../components/feedback/src/FeedbackElement.js";

const Feedback = createComponent({
  tagName: "cds-aichat-feedback",
  elementClass: CDSChatFeedbackElement,
  react: React,
});

export type { FeedbackInitialValues, FeedbackSubmitDetails };
export default Feedback;
