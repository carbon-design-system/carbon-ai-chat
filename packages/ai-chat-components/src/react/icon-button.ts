/*
 *  Copyright IBM Corp. 2025
 *
 *  This source code is licensed under the Apache-2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  @license
 */

import { createComponent } from "@lit/react";
import React from "react";

// Export the actual class for the component that will *directly* be wrapped with React.
import CarbonIconButtonElement from "@carbon/web-components/es/components/icon-button/icon-button.js";

const IconButton = createComponent({
  tagName: "cds-icon-button",
  elementClass: CarbonIconButtonElement,
  react: React,
});

export default IconButton;
