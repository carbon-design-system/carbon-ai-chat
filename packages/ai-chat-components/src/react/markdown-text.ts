/*
 *  Copyright IBM Corp. 2025
 *
 *  This source code is licensed under the Apache-2.0 license found in the
 *  LICENSE file in the root directory of this source tree.
 *
 *  @license
 */

import { createComponent } from "@lit/react";
import React from "react";

// Export the actual class for the component that will *directly* be wrapped with React.
import CDSAIChatMarkdownText from "../components/markdown-text/src/cds-aichat-markdown-text.js";

const MarkdownText = createComponent({
  tagName: "cds-aichat-markdown-text",
  elementClass: CDSAIChatMarkdownText,
  react: React,
});

export default MarkdownText;
