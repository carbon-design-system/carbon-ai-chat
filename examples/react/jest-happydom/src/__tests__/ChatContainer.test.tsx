import React from "react";
import { render, act, waitFor } from "@testing-library/react";
import {
  ChatContainer,
  PageObjectId,
  MessageResponseTypes,
} from "@carbon/ai-chat";

/**
 * These tests demonstrate Jest testing with @carbon/ai-chat React components using happy-dom.
 *
 * IMPORTANT: happy-dom DOES support shadow DOM and Lit components DO render!
 * We can query elements inside the shadow DOM using PageObjectId selectors.
 */
describe("ChatContainer", () => {
  it("should render the chat component", async () => {
    const { container } = await act(() =>
      render(
        <ChatContainer
          messaging={{
            customSendMessage(_request, _requestOptions, instance) {
              // Return a welcome message
              instance.messaging.addMessage({
                output: {
                  generic: [
                    {
                      response_type: MessageResponseTypes.TEXT,
                      text: "Hello! How can I help you today?",
                    },
                  ],
                },
              });
            },
          }}
          data-testid="chat-container"
          renderWriteableElements={{
            headerBottomElement: (
              <div data-testid="custom-header">Custom Header Content</div>
            ),
          }}
        />,
      ),
    );

    // Wait for the custom element to be rendered
    const customElement = await waitFor(() =>
      container.querySelector("cds-aichat-react"),
    );
    expect(customElement).toBeInTheDocument();
  }, 60000);

  it("should render slotted content", async () => {
    const { container } = await act(() =>
      render(
        <ChatContainer
          messaging={{
            customSendMessage(_request, _requestOptions, _instance) {
              console.log("customSendMessage");
            },
          }}
          renderWriteableElements={{
            headerBottomElement: (
              <div data-testid="custom-header">Custom Header Content</div>
            ),
          }}
        />,
      ),
    );

    // This is testing the content passed into the slot, not where it is slotted. This
    // means that even if it is not currently slotted by Carbon AI Chat and is just
    // invisible on the screen, you can still access it for your tests.
    const slotWrapper = await waitFor(() =>
      container.querySelector('[slot="headerBottomElement"]'),
    );
    expect(slotWrapper).toBeInTheDocument();

    const customHeader = slotWrapper?.querySelector(
      '[data-testid="custom-header"]',
    );
    expect(customHeader).toBeInTheDocument();
    expect(customHeader).toHaveTextContent("Custom Header Content");

    await waitFor(() => container.querySelector("cds-aichat-react")).then(() =>
      expect(container.firstChild).toMatchSnapshot(),
    );
  }, 60000);

  it("should have a shadow DOM with rendered content", async () => {
    const { container } = await act(() =>
      render(
        <ChatContainer
          messaging={{
            customSendMessage(_request, _requestOptions, instance) {
              instance.messaging.addMessage({
                output: {
                  generic: [
                    {
                      response_type: MessageResponseTypes.TEXT,
                      text: "Test message",
                    },
                  ],
                },
              });
            },
          }}
        />,
      ),
    );

    const customElement = await waitFor(() =>
      container.querySelector("cds-aichat-react"),
    );

    // Wait for Lit to finish rendering
    if (typeof (customElement as any).updateComplete !== "undefined") {
      await (customElement as any).updateComplete;
    }

    await waitFor(() => container.querySelector("cds-aichat-react")).then(() =>
      expect(container.firstChild).toMatchSnapshot(),
    );
  }, 60000);

  it("should render PageObjectId elements in shadow DOM", async () => {
    const { container } = await act(() =>
      render(
        <ChatContainer
          messaging={{
            customSendMessage(_request, _requestOptions, instance) {
              instance.messaging.addMessage({
                output: {
                  generic: [
                    {
                      response_type: MessageResponseTypes.TEXT,
                      text: "Test message",
                    },
                  ],
                },
              });
            },
          }}
        />,
      ),
    );

    const customElement = await waitFor(() =>
      container.querySelector("cds-aichat-react"),
    );

    // Wait for Lit rendering
    if (typeof (customElement as any).updateComplete !== "undefined") {
      await (customElement as any).updateComplete;
    }

    const shadowRoot = (customElement as any)?.shadowRoot;

    // The LAUNCHER should be present (in minimized state)
    const launcher = shadowRoot.querySelector(
      `[data-testid="${PageObjectId.LAUNCHER}"]`,
    );
    expect(launcher).toBeTruthy();
    await waitFor(() => container.querySelector("cds-aichat-react")).then(() =>
      expect(container.firstChild).toMatchSnapshot(),
    );
  }, 60000);

  it("should open chat when launcher is clicked", async () => {
    const { container } = await act(() =>
      render(
        <ChatContainer
          messaging={{
            customSendMessage(_request, _requestOptions, instance) {
              instance.messaging.addMessage({
                output: {
                  generic: [
                    {
                      response_type: MessageResponseTypes.TEXT,
                      text: "Welcome! How can I help you?",
                    },
                  ],
                },
              });
            },
          }}
        />,
      ),
    );

    const customElement = await waitFor(() =>
      container.querySelector("cds-aichat-react"),
    );

    // Wait for Lit rendering
    if (typeof (customElement as any).updateComplete !== "undefined") {
      await (customElement as any).updateComplete;
    }

    const shadowRoot = (customElement as any)?.shadowRoot;

    // Find the launcher button
    const launcher = shadowRoot.querySelector(
      `[data-testid="${PageObjectId.LAUNCHER}"]`,
    );
    expect(launcher).toBeTruthy();

    // Click the launcher to open the chat
    await act(async () => {
      launcher.click();
      // Wait for next render cycle
      if (typeof (customElement as any).updateComplete !== "undefined") {
        await (customElement as any).updateComplete;
      }
    });

    // Verify main panel is now visible
    const mainPanel = await waitFor(() =>
      shadowRoot.querySelector(`[data-testid="${PageObjectId.MAIN_PANEL}"]`),
    );
    expect(mainPanel).toBeTruthy();

    // Verify input field and send button are present
    const inputField = shadowRoot.querySelector(
      `[data-testid="${PageObjectId.INPUT}"]`,
    );
    expect(inputField).toBeTruthy();

    const sendButton = shadowRoot.querySelector(
      `[data-testid="${PageObjectId.INPUT_SEND}"]`,
    );
    expect(sendButton).toBeTruthy();

    await waitFor(() => container.querySelector("cds-aichat-react")).then(() =>
      expect(container.firstChild).toMatchSnapshot(),
    );
  }, 60000);
});
