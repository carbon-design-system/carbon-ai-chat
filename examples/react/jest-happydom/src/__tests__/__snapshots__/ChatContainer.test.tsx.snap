// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ChatContainer should have a shadow DOM with rendered content 1`] = `
<cds-aichat-react>
  <div
    slot="aiTooltipAfterDescriptionElement"
  />
  <div
    slot="welcomeNodeBeforeElement"
  />
  <div
    slot="headerBottomElement"
  />
  <div
    slot="beforeInputElement"
  />
  <div
    slot="homeScreenHeaderBottomElement"
  />
  <div
    slot="homeScreenAfterStartersElement"
  />
  <div
    slot="homeScreenBeforeInputElement"
  />
  <div
    slot="customPanelElement"
  />
</cds-aichat-react>
`;

exports[`ChatContainer should open chat when launcher is clicked 1`] = `
<cds-aichat-react>
  <div
    slot="aiTooltipAfterDescriptionElement"
  />
  <div
    slot="welcomeNodeBeforeElement"
  />
  <div
    slot="headerBottomElement"
  />
  <div
    slot="beforeInputElement"
  />
  <div
    slot="homeScreenHeaderBottomElement"
  />
  <div
    slot="homeScreenAfterStartersElement"
  />
  <div
    slot="homeScreenBeforeInputElement"
  />
  <div
    slot="customPanelElement"
  />
</cds-aichat-react>
`;

exports[`ChatContainer should render PageObjectId elements in shadow DOM 1`] = `
<cds-aichat-react>
  <div
    slot="aiTooltipAfterDescriptionElement"
  />
  <div
    slot="welcomeNodeBeforeElement"
  />
  <div
    slot="headerBottomElement"
  />
  <div
    slot="beforeInputElement"
  />
  <div
    slot="homeScreenHeaderBottomElement"
  />
  <div
    slot="homeScreenAfterStartersElement"
  />
  <div
    slot="homeScreenBeforeInputElement"
  />
  <div
    slot="customPanelElement"
  />
</cds-aichat-react>
`;

exports[`ChatContainer should render slotted content 1`] = `
<cds-aichat-react>
  <div
    slot="aiTooltipAfterDescriptionElement"
  />
  <div
    slot="welcomeNodeBeforeElement"
  />
  <div
    slot="headerBottomElement"
  >
    <div
      data-testid="custom-header"
    >
      Custom Header Content
    </div>
  </div>
  <div
    slot="beforeInputElement"
  />
  <div
    slot="homeScreenHeaderBottomElement"
  />
  <div
    slot="homeScreenAfterStartersElement"
  />
  <div
    slot="homeScreenBeforeInputElement"
  />
  <div
    slot="customPanelElement"
  />
</cds-aichat-react>
`;
